packages/editor/src/components/post-taxonomies/hierarchical-term-selector.js

```diff
@@ -214,9 +215,22 @@ class HierarchicalTermSelector extends Component {
 
 	renderTerms( renderedTerms ) {
 		const { terms = [] } = this.props;
+		const { filterValue } = this.state;
 		return renderedTerms.map( ( term ) => {
 			const id = `editor-post-taxonomies-hierarchical-term-${ term.id }`;
-			return (
+			const show = (
+				// No filter set.
+				'' === filterValue ||
+				// Filter value appears in the term name.
+				-1 !== term.name.toLowerCase().indexOf( filterValue.toLowerCase() ) ||
+				// Filter value appers in any of the children.
+				( term.children.length > 0 && term.children.map(
+					( childTerm ) => -1 !== childTerm.name.toLowerCase().indexOf( filterValue.toLowerCase() )
+				).reduce(
+					( appeared, appears ) => appeared || appears
+				) )
```

Comment:
Seems all this will work only for children. What about grandchildren and deeper nested terms?